protocol.md

```diff
@@ -0,0 +1,748 @@
+retry
+
+#### Error code `0xFF` -- fatal protocol error
+
+Connection will close after this frame. message ID of this frame should be
+`0xFFFFFFFF`.
+
+### id:4
+
+Message id of the original request that triggered this error, or `0xFFFFFFFF` if
+no message id is available.
+
+### message~2
+
+A human readable string not intended to be shown to the user. This is something
+that goes into error logs to help engineers in the future debug code they've
+never seen before. We will likely need to adopt some kind of convention around
+the contents of this field, particularly so that they can be parseable for
+proper search indexing and aggregation. However, for the purposes of the
+protocol, the contents of "message" are intentionally not specified.
+
+## Transport Headers
+
+These headers are intended to control things at the transport and routing
+layer, so they are expected to be small and inexpensive to process.
+Application level headers can be expressed at a higher level in the protocol.
+
+Duplicate header keys are not allowed and should result in a parse error.
+
+Header keys may not be 0 length (should result in a parse error), but header
+values may be 0 length.
+
+Schema: `nh:1 (hk~1 hv~1){nh}`
+
+TChannel headers are 0 or more UTF-8 string key/value pairs.
+
+The number of headers is sent with the first byte (`nh`).
+
+If `nh` is 0 then there are no bytes in the headers section.
+
+If nh is 1 or more, then that many key/value pairs follow.
+
+Each key and value string is preceded by one byte encoding its length.
+
+For example, the folloing hex dump:
+
+```
+0103 6369 6402 6869  ..cid.hi
+```
+
+Encodes exactly one key/val pair: `("cid", "hi")`.
+
+The following table ilsts the valid transport header keys and whether they are
+valid or not in a call req or res.  Following sections will elaborate on details.
+
+name  | req | res | description
+------|-----|-----|---------------
+`as`  | Y   | Y   | the Arg Scheme
+`cas` | Y   | N   | Claim At Start
+`caf` | Y   | N   | Claim At Finish
+`re`  | Y   | N   | Retry Flags
+`se`  | Y   | N   | Speculative Execution
+`fd`  | Y   | Y   | Failure Domain
+
+### Transport Header `as` -- Arg Scheme
+
+This describes the format of the args for endpoint handlers and/or protocol
+inspectors. The primary RPC mechanism will use "thrift". "http" and "json" are
+used for interop with other systems.
+
+The following table describes the valid `as` values and their corresponding
+arg1, arg2, and arg3 forms:
+
+value    | arg1                  | arg2                                     | arg3
+---------|-----------------------|------------------------------------------|---------------
+`thrift` | string method name    | application headers `nh:2 (k~2 v~2){nh}` | thrift payload
+`json`   | method name as string | application headers as JSON              | body as JSON
+`http`   | method + URI          | headers as JSON array                    | raw body
```

Comment:
It seems kind of weird to me that we spend a bunch of time making short header names and single-byte codes to define their values, but then go full-on string for `as`. I like the readability, but it feels inconsistent. Perhaps we should consider an `as` code table?